go-controller/pkg/ovn/address_set_sync.go

```diff
@@ -0,0 +1,447 @@
+	// staleAS will be updated if there is no newer version in newASs, and deleted if there is
+	staleASOps map[string]staleASInfo
+	newASs     map[string]bool
+	// one referencing object may contain multiple references that need to be updated
+	// these maps are used to track referenced that need to be replace for every object type
+	replaceRefACLs  map[string]refsList[*nbdb.ACL]
+	replaceRefQoSes map[string]refsList[*nbdb.QoS]
+	replaceRefLRPs  map[string]refsList[*nbdb.LogicalRouterPolicy]
+}
+
+// controllerName is the name of the new controller that should own all address sets without controller
+func newAddressSetSyncer(nbClient libovsdbclient.Client, controllerName string) *addressSetsSyncer {
+	return &addressSetsSyncer{
+		nbClient:        nbClient,
+		controllerName:  controllerName,
+		txnBatchSize:    50,
+		staleASOps:      map[string]staleASInfo{},
+		newASs:          map[string]bool{},
+		replaceRefACLs:  map[string]refsList[*nbdb.ACL]{},
+		replaceRefQoSes: map[string]refsList[*nbdb.QoS]{},
+		replaceRefLRPs:  map[string]refsList[*nbdb.LogicalRouterPolicy]{},
+	}
+}
+
+// return if address set is owned by network policy and its namespace, name, internal id
+func checkIfNetpol(asName string) (netpolOwned bool, namespace, name, direction, idx string) {
+	// old format fmt.Sprintf("%s.%s.%s.%d", gp.policyNamespace, gp.policyName, direction, gp.idx)
+	// namespace doesn't have dots
+	s := strings.Split(asName, ".")
+	sLen := len(s)
+	// index should be a number
+	_, numErr := strconv.Atoi(s[sLen-1])
+	if sLen >= 4 && (s[sLen-2] == "ingress" || s[sLen-2] == "egress") && numErr == nil {
+		// address set is owned by network policy
+		netpolOwned = true
+		// namespace doesn't have dots
+		namespace = s[0]
+		// policyName may have dots, join in that case
+		name = strings.Join(s[1:sLen-2], ".")
+		direction = s[sLen-2]
+		idx = s[sLen-1]
+	}
+	return
+}
+
+// getReferencingObjsAndNewDbIDs finds all object that reference stale address set and tries to create a new dbIDs
+// based on referencing objects
+func (syncer *addressSetsSyncer) getReferencingObjsAndNewDbIDs(oldHash, oldName string) (acls []*nbdb.ACL,
+	qoses []*nbdb.QoS, lrps []*nbdb.LogicalRouterPolicy, dbIDs *libovsdbops.DbObjectIDs, err error) {
+	// find all referencing objects
+	aclPred := func(acl *nbdb.ACL) bool {
+		return strings.Contains(acl.Match, "$"+oldHash)
+	}
+	acls, err = libovsdbops.FindACLsWithPredicate(syncer.nbClient, aclPred)
+	if err != nil {
+		err = fmt.Errorf("failed to find acls for address set %s: %v", oldHash, err)
+		return
+	}
+	qosPred := func(qos *nbdb.QoS) bool {
+		return strings.Contains(qos.Match, "$"+oldHash)
+	}
+	qoses, err = libovsdbops.FindQoSesWithPredicate(syncer.nbClient, qosPred)
+	if err != nil {
+		err = fmt.Errorf("failed to find qoses for address set %s: %v", oldHash, err)
+		return
+	}
+	lrpPred := func(lrp *nbdb.LogicalRouterPolicy) bool {
+		return strings.Contains(lrp.Match, "$"+oldHash)
```

Comment:
Anything else we can match on to make sure this are `ovntypes.OVNClusterRouter` policies? I am bit concerned this changes somewhere down the road.

Reply:
good point, will add a check

Reply:
Hmm is there something I am missing here or would have it been a (probably better) alternative to have an update only method (instead of CreateOrUpdate) for the LRP.
I suggested if it was viable to do it in the predicate thinking there could probably be an easy way based on the LRP itself.